contracts/utils/math/Math.sol

```diff
@@ -335,6 +335,55 @@ library Math {
         }
     }
 
+    /**
+     * @dev Variant of {modExp} that supports inputs of arbitrary length.
+     */
+    function modExp(bytes memory b, bytes memory e, bytes memory m) internal view returns (bytes memory) {
+        (bool success, bytes memory result) = tryModExp(b, e, m);
+        if (!success) {
+            if (_zeroArray(m)) {
+                Panic.panic(Panic.DIVISION_BY_ZERO);
+            } else {
+                revert Address.FailedInnerCall();
+            }
+        }
+        return result;
+    }
+
+    /**
+     * @dev Variant of {tryModExp} that supports inputs of arbitrary length.
+     */
+    function tryModExp(
+        bytes memory b,
+        bytes memory e,
+        bytes memory m
+    ) internal view returns (bool success, bytes memory result) {
+        if (_zeroArray(m)) return (false, new bytes(0));
+
+        // Encode call args and move the free memory pointer
+        bytes memory args = abi.encodePacked(b.length, e.length, m.length, b, e, m);
+
+        // Given result <= modulus
+        result = new bytes(m.length);
+
+        /// @solidity memory-safe-assembly
+        assembly {
+            success := staticcall(gas(), 0x05, add(args, 0x20), mload(args), add(result, 0x20), mload(m))
+        }
```

Comment:
This is going to leak more memory than needed.

- `args` are encoded, on new memory, that is allocate at the free memory pointer (0x40)
- then `result` is allocated (after args)
- in the end, `result` is returned, and `args` is never dealocated.

This means that if you do that in a loop, your memory pointer moves, and the gas cost increase quadratically.

I did implement a version that leaks less memory (only the result array is allocated) [here](https://github.com/Amxx/openzeppelin-contracts/blob/feature/math/modExp-bytes/contracts/utils/math/Math.sol#L356-L381). Any reason to not use that version?

Reply:
This is becoming a guideline. Generally we should consider memory leaking in almost everything.

I'm out of home rn but will update once back. Thanks for flagging!